specification/sections/02-providers.md

```diff
@@ -150,3 +151,85 @@ class MyProvider implements Provider {
 #### Requirement 2.3.3
 
 > In cases of abnormal execution, the `resolution details` structure's `error message` field **MAY** contain a string containing additional detail about the nature of the error.
+
+### 2.4 Initialization
+
+[![experimental](https://img.shields.io/static/v1?label=Status&message=experimental&color=orange)](https://github.com/open-feature/spec/tree/main/specification#experimental)
+
+#### Requirement 2.4.1
+
+> The `provider` interface **MUST** define an optional `initialize` function, which accepts the global `evaluation context` as an argument and performs initialization logic relevant to the provider.
+
+Many feature flag frameworks or SDKs require some initialization before they can be used.
+They might require the completion of an HTTP request, establishing persistent connections, or starting timers or threads.
+The `initialization` function is an ideal place for such logic.
+
+```java
+// MyProvider implementation of the initialize function defined in Provider
+class MyProvider implements Provider {
+  //...
+
+  // the global context is passed to the initialization function
+  void initialize(EvaluationContext initialContext) {
+    /*
+      A hypothetical initialization function: connect to a flag database, then pass a callback which fires PROVIDER_READY
+    */
+    this.connectToDatabase(() => this.emitEvent(ProviderEvents.READY));
+  }
+
+  //...
+}
+```
+
+### 2.5. Provider context reconciliation
+
+[![experimental](https://img.shields.io/static/v1?label=Status&message=experimental&color=orange)](https://github.com/open-feature/spec/tree/main/specification#experimental)
+
+Static-context focused providers may need a mechanism to understand when their cache of evaluated flags must be invalidated or updated. An `on-context-set` handler can be defined which performs whatever operations are needed to reconcile the evaluated flags with the new context.
+
+#### Requirement 2.5.1
+
+> The provider interface **MUST** define an `on context set` handler, which takes an argument for the previous context, and the newly set context, and which can be optionally implemented to reconcile any stored state pertaining to the global evaluation context.
```

Comment:
There's a lot of overly specifying about what the interface must be. This kinda feels wrong to me. Not all languages have interfaces. I would prefer that we talk about what providers may do, and what the sdk must do if they do that.

Instead of saying "the provider interface must define a handler".. I'd rather this read: "Providers may register a context change handler. Separately, if the context changes, the sdk must call that handler".

Reply:
Thinking about it more, there's a few places where I've used **MUST** to describe optional fields/methods. I think generally, that you're right, and this can be converted to **MAY**, mentioning the specific constrcut. For example

`The provider interface MUST have optional field X which does Y` 

becomes:

`The provider MAY have X which does Y`.

I think this is generally better and and I've applied it in a few places.